File: scripts/mrp_pipeline/models/atomic_node.py
from typing import List, Optional, Dict
from dataclasses import dataclass, field


@dataclass
class CausalWeb:
    """Các mối quan hệ nhân quả trong nốt nguyên tử."""
    causal_core: Optional[str] = None          # Nhân gốc (slug của parent/ancestor concept)
    supporting_conditions: List[str] = field(default_factory=list) # Hội tụ Duyên (other concepts)
    derivative_effects: List[str] = field(default_factory=list)    # Quả chuyển hóa (concepts created because of this)


@dataclass
class AtomicNode:
    """
    Đại diện cho một nốt tri thức nguyên tử (Layer 3) ở 02_atomic_nodes/.
    Đảm bảo tính đơn nhất (Singularity) và cấu trúc liên kết đồ thị phẳng.
    """
    slug: str                     # Tên file không prefix/suffix (vd: "context-manager")
    title: str                    # Tên khái niệm rõ ràng, sắc bén
    category: str                 # Danh mục phân loại
    tags: List[str]               # Thẻ phân loại
    definition: str               # Định nghĩa & Nội dung Cốt lõi
    principles: List[str]         # Nguyên lý Kỹ thuật & Thực tiễn
    parent: Optional[str] = None  # Slug của Node cha
    children: List[str] = field(default_factory=list) # Danh sách slug của Node con
    causal_web: CausalWeb = field(default_factory=CausalWeb) # Mối liên hệ nhân quả
    evidence_structured: List[str] = field(default_factory=list)  # Các file processed làm dẫn chứng
    evidence_raw: List[str] = field(default_factory=list)         # Các file thô làm dẫn chứng

    @property
    def full_slug(self) -> str:
        """Slug đầy đủ có prefix (vd: HAE-concept-context-manager)."""
        return f"HAE-concept-{self.slug}"

    @property
    def relative_path(self) -> str:
        """Đường dẫn workspace-relative."""
        return f"02_atomic_nodes/{self.full_slug}.md"

    def to_markdown(self) -> str:
        """Biên dịch nốt sang file Markdown theo đúng hiến pháp Layer 3 (Causal Web & backlinks)."""
        tags_str = "\n".join([f"  - {t}" for t in self.tags])
        principles_str = "\n".join([f"- **{p.split(':')[0]}:** {p.split(':')[1]}" if ":" in p else f"- {p}" for p in self.principles])

        # Phục vụ Causal Web
        core_link = f"[{self.causal_web.causal_core.replace('-', ' ').title()}](02_atomic_nodes/HAE-concept-{self.causal_web.causal_core}.md)" if self.causal_web.causal_core else "Không có"

        duyen_links = []
        for d in self.causal_web.supporting_conditions:
            duyen_links.append(f"[{d.replace('-', ' ').title()}](02_atomic_nodes/HAE-concept-{d}.md)")
        duyen_str = ", ".join(duyen_links) if duyen_links else "Không có"

        qua_links = []
        for q in self.causal_web.derivative_effects:
            qua_links.append(f"[{q.replace('-', ' ').title()}](02_atomic_nodes/HAE-concept-{q}.md)")
        qua_str = ", ".join(qua_links) if qua_links else "Không có"

        # Phục vụ parent/children
        parent_fm = f"\nparent: {self.parent}" if self.parent else ""
        children_fm = ""
        if self.children:
            children_fm = "\nchildren:\n" + "\n".join([f"  - {c}" for c in self.children])

        # Dẫn chứng ngược dòng
        evidence_str = ""
        for s in self.evidence_structured:
            evidence_str += f"  - [Ghi chú cấu trúc: {s.replace('-processed', '').replace('-', ' ').title()}](01_structured_docs/{s}.md)\n"
        for r in self.evidence_raw:
            evidence_str += f"  - [Ghi chú thô: {r.replace('-', ' ').title()}](00_raw_docs/{r}.md)\n"

        # Con nếu có
        subnodes_section = ""
        if self.children:
            subnodes_links = "\n".join([f"- [{c.replace('-', ' ').title()}](02_atomic_nodes/HAE-concept-{c}.md)" for c in self.children])
            subnodes_section = f"""
## 🌳 Nốt con (Sub-Nodes)
{subnodes_links}
"""

        content = f"""---
id: {self.full_slug}
title: "{self.title}"
category: "{self.category}"
tags:
{tags_str}
date: 2026-05-31{parent_fm}{children_fm}
---

# {self.title}

## 💡 Định nghĩa & Nội dung Cốt lõi
{self.definition}

## ⚙️ Nguyên lý Kỹ thuật & Thực tiễn
{principles_str}
{subnodes_section}
## 🔗 Mạng lưới Nhân Duyên Quả (Causal Web)
- **Nhân gốc (Causal Core)**: {core_link} — Khái niệm nền tảng sinh ra khái niệm này.
- **Hội tụ Duyên (Supporting Conditions)**: {duyen_str} — Các khái niệm hỗ trợ trực tiếp.
- **Quả chuyển hóa (Derivative Effects)**: {qua_str} — Các giải pháp và hiệu quả kế thừa.

## 🔗 Liên kết Tri thức (Knowledge Connections)
- **Dẫn chứng & Nguồn gốc (Ngược dòng - Evidence & Context)**:
{evidence_str.rstrip()}
- **Đúc kết vĩ mô (Xuôi dòng - Distilled Thoughts)**:
  - [Tuyên ngôn Harness Engineering](04_distilled/harness-engineering-manifesto.md)
"""
        return content

File: src/models/atomic-node.ts
import { ATOMIC_PREFIX } from '../core/config.ts';

export interface CausalWeb {
  causalCore?: string;             // Nhân gốc (slug của parent/ancestor concept)
  supportingConditions: string[]; // Hội tụ Duyên (other concepts)
  derivativeEffects: string[];    // Quả chuyển hóa (concepts created because of this)
}

export class AtomicNode {
  slug: string;                     // Tên file không prefix/suffix (vd: "context-manager")
  title: string;                    // Tên khái niệm rõ ràng, sắc bén
  category: string;                 // Danh mục phân loại
  tags: string[];                   // Thẻ phân loại
  definition: string;               // Định nghĩa & Nội dung Cốt lõi
  principles: string[];             // Nguyên lý Kỹ thuật & Thực tiễn
  parent?: string;                  // Slug của Node cha
  children: string[];               // Danh sách slug của Node con
  causalWeb: CausalWeb;             // Mối liên hệ nhân quả
  evidenceStructured: string[];     // Các file processed làm dẫn chứng
  evidenceRaw: string[];            // Các file thô làm dẫn chứng

  constructor(data: {
    slug: string;
    title: string;
    category: string;
    tags: string[];
    definition: string;
    principles: string[];
    parent?: string;
    children?: string[];
    causalWeb?: CausalWeb;
    evidenceStructured?: string[];
    evidenceRaw?: string[];
  }) {
    this.slug = data.slug;
    this.title = data.title;
    this.category = data.category;
    this.tags = data.tags;
    this.definition = data.definition;
    this.principles = data.principles;
    this.parent = data.parent;
    this.children = data.children || [];
    this.causalWeb = data.causalWeb || { supportingConditions: [], derivativeEffects: [] };
    this.evidenceStructured = data.evidenceStructured || [];
    this.evidenceRaw = data.evidenceRaw || [];
  }

  get fullSlug(): string {
    return `${ATOMIC_PREFIX}${this.slug}`;
  }

  get relativePath(): string {
    return `02_atomic_nodes/${this.fullSlug}.md`;
  }

  toMarkdown(): string {
    const tagsStr = this.tags.map(t => `  - ${t}`).join('\n');
    const principlesStr = this.principles
      .map(p => {
        if (p.includes(':')) {
          const parts = p.split(':');
          const prefix = parts[0];
          const suffix = parts.slice(1).join(':');
          return `- **${prefix.trim()}:** ${suffix.trim()}`;
        }
        return `- ${p}`;
      })
      .join('\n');

    // Phục vụ Causal Web
    const coreLink = this.causalWeb.causalCore
      ? `[${this.causalWeb.causalCore.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}](02_atomic_nodes/${ATOMIC_PREFIX}${this.causalWeb.causalCore}.md)`
      : 'Không có';

    const duyenLinks = this.causalWeb.supportingConditions.map(
      d => `[${d.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}](02_atomic_nodes/${ATOMIC_PREFIX}${d}.md)`
    );
    const duyenStr = duyenLinks.length > 0 ? duyenLinks.join(', ') : 'Không có';

    const quaLinks = this.causalWeb.derivativeEffects.map(
      q => `[${q.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}](02_atomic_nodes/${ATOMIC_PREFIX}${q}.md)`
    );
    const quaStr = quaLinks.length > 0 ? quaLinks.join(', ') : 'Không có';

    // Phục vụ parent/children
    const parentFm = this.parent ? `\nparent: ${this.parent}` : '';
    let childrenFm = '';
    if (this.children.length > 0) {
      childrenFm = '\nchildren:\n' + this.children.map(c => `  - ${c}`).join('\n');
    }

    // Dẫn chứng ngược dòng
    let evidenceStr = '';
    for (const s of this.evidenceStructured) {
      const title = s.replace('-processed', '').replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase());
      evidenceStr += `  - [Ghi chú cấu trúc: ${title}](01_structured_docs/${s}.md)\n`;
    }
    for (const r of this.evidenceRaw) {
      const title = r.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase());
      evidenceStr += `  - [Ghi chú thô: ${title}](00_raw_docs/${r}.md)\n`;
    }

    // Con nếu có
    let subnodesSection = '';
    if (this.children.length > 0) {
      const subnodesLinks = this.children
        .map(c => `- [${c.replace(/-/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}](02_atomic_nodes/${ATOMIC_PREFIX}${c}.md)`)
        .join('\n');
      subnodesSection = `\n## 🌳 Nốt con (Sub-Nodes)\n${subnodesLinks}\n`;
    }

    return `---
id: ${this.fullSlug}
title: "${this.title}"
category: "${this.category}"
tags:
${tagsStr}
date: 2026-05-31${parentFm}${childrenFm}
---

# ${this.title}

## 💡 Định nghĩa & Nội dung Cốt lõi
${this.definition}

## ⚙️ Nguyên lý Kỹ thuật & Thực tiễn
${principlesStr}
${subnodesSection}
## 🔗 Mạng lưới Nhân Duyên Quả (Causal Web)
- **Nhân gốc (Causal Core)**: ${coreLink} — Khái niệm nền tảng sinh ra khái niệm này.
- **Hội tụ Duyên (Supporting Conditions)**: ${duyenStr} — Các khái niệm hỗ trợ trực tiếp.
- **Quả chuyển hóa (Derivative Effects)**: ${quaStr} — Các giải pháp và hiệu quả kế thừa.

## 🔗 Liên kết Tri thức (Knowledge Connections)
- **Dẫn chứng & Nguồn gốc (Ngược dòng - Evidence & Context)**:
${evidenceStr.trimEnd()}
- **Đúc kết vĩ mô (Xuôi dòng - Distilled Thoughts)**:
  - [Đúc kết Kinh điển Nikaya](04_distilled/nikaya-distilled.md)
`;
  }
}
